src/ImageSharp/ColorProfiles/Icc/Calculators/LutEntryCalculator.cs

```diff
@@ -37,7 +37,7 @@ public Vector4 Calculate(Vector4 value)
     {
         if (this.doTransform)
         {
-            value = Vector4.Transform(value, this.matrix);
+            value = Vector4.Transform(value, Matrix4x4.Transpose(this.matrix));
```

Comment:
It seems `Vector4.Transform` was performing _Vector × Matrix_ not  _Matrix × Vector_? My matrix maths is rusty, but in lieu of some kind of `Matrix4x4.Transform(matrix, vector)` function it sounds like transposing the matrix has the same effect as changing the order of multiplication.

This issue wasn't caught yet because none of the ICC conversion tests use a profile of CMYK data ⇒ XYZ PCS. And even if it did, it would only get caught if using a non-identity matrix? As always, I'll be surprised if we can find a profile that meets these conditions, though we can include it in testing if we do.

Reply:
Honestly, I'm not sure. @saucecontrol could I borrow your brain here please?

Reply:
Vector4.Transform is a dot product, which is commutative between a vector and matrix.

I can work up some profiles to test some of these edge cases if you'd like.  I swear one of these days I'm going to get around to finishing my implementation of LUT profile transforms, and I'll need them then anyway.

Reply:
Here we want a matrix multiplication M×V which isn't commutative, but doing V⋅M<sup>T</sup> effectively calculates M×V in this particular case.

Is that an OK assumption to make given we're restricted to `Matrix4x4` and `Vector4`? Is there a more suitable `System.Numerics` way to do this?

Reply:
You don't want M×V, you want M⋅V, and I promise [it's commutative](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs,647).

The reason transposing gives you the correct result is that `Vector4.Transpose` assumes the matrix is row-major, i.e. the V is a 1x4 matrix.  Most of ICC docs describe the matrices as column-major, which would mean your V is a 4x1 matrix.

Reply:
> You don't want M×V, you want M⋅V

I think some of my assumptions have muddied the waters a bit, so I'm going to try to summarise, and if I'm still making a mess I'll have to leave this to someone else.
- Until now I've basically ignored that the vector is a 1x4 matrix, and I'm sure I've confused things by using V to mean "the values of the vector" as opposed to "a 1x4 matrix"
- In terms of the maths I want to perform, I want to multiply a 4x4 matrix with a 4x1 matrix (what I was referring to as not commutative), and there's no in-built function for that
- In terms of the implementation, it sounds like we want to perform a dot product of a 4x4 matrix with a 1x4 matrix (which is commutative) for SIMD reasons
- Since I'm stuck with a row-major 1x4 matrix, I need to make the 4x4 matrix row-major for the dot product calculation
- I'm not experienced enough with `System.Numerics` to know if there's a more conventional way to do the matrix multiplication than transposing the 4x4 matrix

Reply:
> In terms of the maths I want to perform, I want to multiply a 4x4 matrix with a 4x1 matrix (what I was referring to as not commutative), and there's no in-built function for that?

This is also commutative, as it is also a dot product.  The only difference is whether you compute the dot product of the vector with each row of the matrix or with each column of the matrix.

There is no fast way to compute it column-wise, because it means gathering the non-adjacent values in memory.  In fact, transposing the matrix with SIMD and then performing a row-wise dot product with SIMD is the fast way to do this, although it is *much* less fast than normalizing your matrix to row-major when reading the profile and skipping the transpose step during processing.

> perhaps any 4x4 matrices just need to be initialised row-major?

Exactly

Reply:
Just to be sure the maths explanation is clear:

M⋅V where M is 4x4 can be summarized as VMatrix<sub>1</sub>⋅VColor + VMatrix<sub>2</sub>⋅VColor + VMatrix<sub>3</sub>⋅VColor + VMatrix<sub>4</sub>⋅VColor where VMatrix represents either the rows or columns of the matrix.  That result is the same as VColor⋅VMatrix<sub>1</sub> + VColor⋅VMatrix<sub>2</sub> + VColor⋅VMatrix<sub>3</sub> + VColor⋅VMatrix<sub>4</sub>, because V⋅V is commutative.

Reply:
Sure, I'm referring to matrix multiplication in general.

I'm not sure how feasible initialising row-major is when we also need the inverse of the matrix for reverse transforms, will see.